README.md
# TA-nmon


docs/releasenotes.rst
#########################################
Release notes
#########################################

^^^^^^^^^^^^
Requirements
^^^^^^^^^^^^

* Splunk 6.x / Universal Forwarder v6.x and later Only

* Universal Forwarders clients system lacking a Python 2.7.x interpreter requires Perl WITH Time::HiRes module available

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
What has been fixed by release
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

========
V1.3.33:
========

- fix: review sourcetypes definition #60
- Feature: mutex implementation to avoid simultaneous run of shell scripts #59
- Feature: override serial number #58
- fix: Removing crcSalt from inputs since it is not required anymore #57
- fix: Typo in nmon_helper.sh during Perl / Python detection #56
- fix: Host override feature improvement for non data sourcetypes #55

========
V1.3.32:
========

- fix: batch mode and fishbuckets performance issues #53
- fix: Missing rotation of external header in fifo_reader.py/pl #54

========
V1.3.31:
========

- feature: Solaris - SARMON upgrade to v1.12 (Sparc FIFO mode) #51
- fix: Rename eventgen.conf to avoid splunkd WARN messages #52

========
V1.3.30:
========

- fix: reactivating the JFSFILE / JFSINODE collections until new core release is available to prevent missing features

========
V1.3.29:
========

- fix: Python parser - header detection correction for nmon external monitoring
- fix: unexpected operator issue during process identification #48
- fix: prevent bundle validation warn messages with spec files in README directory
- feature: Add df information for improved file system monitoring and storage capacity planning
- feature: JFSFILE/JFSINODE are being replaced (and deactivated) by external collection with DF_STORAGE/DF_INODES

========
V1.3.28:
========

- fix: Perl parser write new line after each write (Case #508792 Splunk universal forwarders duplicating data due to File too small to check seekcrc #47)
- fix: minor code cleaning
- feature: complete review of eventgen configuration using TA-nmon-hec samples

========
V1.3.27:
========

- feature: index and search time parsing configuration for the nmon-logger-splunk-hec package (agent less package for Splunk http input)
- fix: Fully Qualified Domain Name improvements #46

========
V1.3.26:
========

- fix: AIX - Better management of compatibility issue with topas-nmon not supporting the -y option #43
- fix: AIX - fix repeated and not justified pid file removal message #44
- fix: ALL OS - nmon_helper.sh code improvements #45

========
V1.3.25:
========

- feature: Optimize nmon_processing output and reduce volume of data to be generated #37
- fix: Linux issue: detection of default/nmon.conf rewrite required is incorrect #41
- fix: Error in nmon_helper.sh - bad analysis of external snap processes #42

========
V1.3.24:
========

- fix: AIX issue - non terminating processes will result in collecting stop #38
- fix: time zone issue - additional sourcetypes (collect, clean) use a date format that includes TZ and can lead to confusion #39
- fix: Review interpreter (python/perl) choice in all shell scripts, specially to fix some AIX issues #40

========
V1.3.23:
========

- unpublished intermediate release

========
V1.3.22:
========

- fix: nmon from syslog - missing indexed time creation for OStype and type fields #31
- fix: nmon from syslog - uptime extraction failure #32
- fix: nmon external - manager header in a dedicated file #33
- fix: Perl parser - prevent error message when --json_output option is set #34
- feature: Linux nmon binaries upgrade to nmon 16g #35
- fix: AIX - multiplication of nmon external snap instances on some systems #36

========
V1.3.21:
========

- fix: AIX issue - fifo_reader regex match some monitors into header #29
- fix: Nmon external issue - manage metrics collection into a dedicated file #30

========
V1.3.20:
========

- unpublished intermediate release

========
V1.3.19:
========

- fix: fifo mode implementation in parsers and several corrections #27
- fix: CIM compliance improvements and corrections
- fix: missing oshost tag for ITSI
- fix: fifo_consumer.sh error in file naming for rotated files purge
- fix: fifo_consumer.sh issue generates gaps in data #28
- feature: Allows deactivating fifo mode and switch to old mechanism via nmon.conf #26
- feature: Allows deactivating nmon external generation via nmon.conf #25

==============
V1.3.16 to 18:
==============

- unpublished intermediate releases

========
V1.3.15:
========

- Fix: nmon external load average extraction failure on some OS
- Fix: TA-nmon local/nmon.conf from the SHC deployer is not compatible #23
- Fix: Use the nmon var directory for fifo_consumer.sh temp file management
- Fix: solve nmon_external issues with AIX 6.1/7.1 (collection randomly stops)
- Fix: manage old topas-nmon version not compatible with -y option
- Feature: binaries for Ubuntu 17 (x86 32/64, power)

========
V1.3.14:
========

- intermediate version not published

========
V1.3.13:
========

**This is a major release of the TA-nmon:**

- Feature: fantastic reduction of the system foot print (CPU,I/O,memory) with the new fifo implementation, the TA-nmon cost is now minimal!
- Feature: easily extend the native nmon data with any external data (OS commands, scripts of any kind, shell, perl, python...) in 2 lines of codes
- Feature: easily customize the list of performance monitors to be parsed (using the nmonparser_config.json)
- Feature: choose between legacy csv and json data generation (limited to Python compatible hosts), you can now choose to generate performance data in json format and prioritize storage over performance and licensing volume
- Feature: new dedicated documentation for the TA-nmon, https://readthedocs.org/projects/ta-nmon
- Feature: nmon binaries for Amazon Linux (AMI)
- Fix: Removal of recursive stanza in inputs.conf #21
- Fix: Increase the interval for nmon_cleaning #18

create_agent.py
#!/usr/bin/env python

# Program name: create_agent.py
# Compatibility: Python 2x
# Purpose - Create a customized version of the TA-nmon for the Splunk Nmon App, see https://apps.splunk.com/app/1753
# Author - Guilhem Marchand
# Disclaimer: Distributed on an "AS IS" basis
# Date of first publication - April 2015

# Licence:

# Copyright 2014 Guilhem Marchand

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

# http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Releases Notes:

# - April 2015, V1.0.0: Guilhem Marchand, Initial version
# - 2015/07/12, V1.0.02: Guilhem Marchand
#                      - Default configuration uses now the nmon2csv.sh wrapper which requires update of the script
# - 2015/07/28, V1.0.03: Guilhem Marchand
#                      - Use the new shell wrapper when no option is provided, or if it is explicitely provided
# - 2016/02/15, V1.0.04: Guilhem Marchand: Prevent Linux and Sarmon binaries from being modified
# - 2016/07/31, V1.0.05: Guilhem Marchand: TA-nmon is now named as tgz extension
# - 2016/08/06, V1.0.06: Guilhem Marchand: With the release of the TA-nmon_selfmode, prevent from extract it
# - 2017/03/01, V1.0.07: Guilhem Marchand: Add the --json_output option
# - 2017/03/15, V1.0.08: Guilhem Marchand: Manage options in nmon.conf
# - 2017/05/10, V1.0.09: Guilhem Marchand: Use the TA-nmon tgz archive instead of full nmon app core tgz archive

import sys
import os
import tarfile
import glob
import fnmatch
import argparse
import shutil

version = '1.0.09'

# ###################################################################
#############           Arguments Parser
####################################################################

# Define Arguments

parser = argparse.ArgumentParser()

parser.add_argument('-f', action='store', dest='INFILE',
                    help='Name of the TA-nmon tgz Archive file')

parser.add_argument('--indexname', action='store', dest='INDEX_NAME',
                    help='Customize the Application Index Name (default: nmon)')

parser.add_argument('--agentname', action='store', dest='TA_NMON',
                    help='Define the TA Agent name and root directory')

parser.add_argument('--agentmode', action='store', dest='agentmode',
                    help='Define the Data Processing mode, valid values are: shell,python,perl / Default'
                         ' value is shell')

parser.add_argument('--parsingmode', action='store', dest='parsingmode',
                    help='Define the Data Processing mode for performance data, valid values are: csv,json / Default'
                         ' value is csv')

parser.add_argument('--version', action='version', version='%(prog)s ' + version)

parser.add_argument('--debug', dest='debug', action='store_true')

parser.set_defaults(debug=False)

args = parser.parse_args()

# Set debug
if args.debug:
    debug = True

####################################################################
#############           Functions
####################################################################

# String replacement function
# Can be called by:
# findreplace(path, string_to_search, replace_by, file_extension)

def findreplace(directory, find, replace, filepattern):
    for path, dirs, files in os.walk(os.path.abspath(directory)):
        for filename in fnmatch.filter(files, filepattern):
            filepath = os.path.join(path, filename)

            # Prevents binaries modification
            if "bin/linux" in filepath:
                if debug:
                    print("file " + str(filename) + " is binary or binary related")
            elif "bin/sarmon" in filepath:
                if debug:
                    print("file " + str(filename) + " is binary or binary related")
            else:
                with open(filepath) as f:
                    s = f.read()
                s = s.replace(find, replace)
                with open(filepath, "w") as f:
                    f.write(s)


####################################################################
#############           Main Program
####################################################################

# Check Arguments
if len(sys.argv) < 2:
    print "\n%s" % os.path.basename(sys.argv[0])
    print "\nThis utility had been designed to allow creating customized agents for the Nmon Splunk Application," \
          " please follow these instructions:\n"
    print "- Download the current release of the TA-nmon App in Splunk Base: https://splunkbase.splunk.com/app/3248 or clone the git repository: https://github.com/guilhemmarchand/TA-nmon.git"
    print "- Ensure to have this Python script and the TGZ archive in the same directory"
    print "- Run the tool: ./create_agent.py and check for available options"
    print "- After the execution, a new agent package will have been created in the resources directory"
    print "- Extract its content to your Splunk deployment server, configure the server class, associated clients and" \
          " deploy the agent"
    print "- Don't forget to set the application to restart splunkd after deployment\n"
    sys.exit(0)

# Will expect in first Argument the name of the tgz Archive of the Application to be downloaded in Splunk Base
if not (args.INFILE):
    print "\nERROR: Please provide the tgz Archive file with -f statement\n"
    sys.exit(1)
else:
    infile = args.INFILE

# Will expect the customize name of the Splunk index, the default name of the index is "nmon"
if not (args.INDEX_NAME):
    print "INFO: No custom index name were provided, using default \"nmon\" name for index"
    index_name = "nmon"
else:
    index_name = args.INDEX_NAME

# Define Agent mode
if not (args.agentmode):
    agentmode = "shell"
else:
    agentmode = args.agentmode

# Define Parsing mode
if not (args.parsingmode):
    parsingmode = "csv"
else:
    parsingmode = args.parsingmode

# If the root directory of the TA-nmon is not defined, exit and show message
if not (args.TA_NMON):
    print "ERROR: You must specify the name of the agent package you want to create, and it must be different from" \
          " the default package TA-nmon"
    sys.exit(0)
else:
    ta_root_dir = args.TA_NMON

# Avoid naming the TA as nmon core application
if ta_root_dir == "nmon":
    print "ERROR: The TA package cannot have the name nmon to avoid collision with the core application configuration"
    sys.exit(1)

# Verify tgz Archive file exists
if not os.path.exists(infile):
    print ('ERROR: invalid file, could not find: ' + infile)
    sys.exit(1)

# Ensure the same package name does not already exist in current directory
if os.path.exists(ta_root_dir):
    print ('ERROR: A directory named ' + ta_root_dir + ' already exist in current directory, please remove it and'
                                                       ' restart')
    sys.exit(1)
elif os.path.exists(ta_root_dir + ".tgz"):
    print ('ERROR: A tgz archive named ' + ta_root_dir + ".tgz" + ' already exist in current directory, please'
                                                                  ' remove it and restart')
    sys.exit(1)

# Extract Archive
tar = tarfile.open(infile)
msg = 'Extracting tgz Archive: ' + infile
print (msg)
tar.extractall()
tar.close()

# Operate

# Get current directory
curdir = os.getcwd()

# Extract the TA-nmon default package in current directory

print ('INFO: Extracting Agent tgz resources Archives')

tgz_files = 'TA-nmon*.tgz'

for tgz in glob.glob(str(tgz_files)):

    # Don't manage the alternative TA-nmon_selfmode
    if not "TA-nmon_selfmode" in tgz:
        tar = tarfile.open(tgz)
        tar.extractall()
        tar.close()

# Rename the TA-nmon directory to match agent name

msg = 'INFO: Renaming TA-nmon default agent to ' + ta_root_dir
print (msg)

shutil.copytree('TA-nmon', ta_root_dir)

################# STRING REPLACEMENTS #################

# Replace the old agent name in files

# Achieve string replacements

print ('Achieving files transformation...')

search = 'TA-nmon'
replace = ta_root_dir
findreplace(ta_root_dir, search, replace, "*.sh")
findreplace(ta_root_dir, search, replace, "*.py")
findreplace(ta_root_dir, search, replace, "*.pl")
findreplace(ta_root_dir, search, replace, "*.conf")

print ('Done.')

# Change index name only if differs from default index name (nmon)
if index_name != "nmon":

    # Replace basic index calls
    print ('INFO: Customizing any reference to index name in files')

    search = 'index = nmon'
    replace = 'index = ' + index_name
    findreplace(ta_root_dir, search, replace, "*.conf")

################# AGENT MODE #################

# shell is default value, if set to python or perl create a custom local/props.conf with appropriate settings

if agentmode in ('python', 'perl'):
    os.mkdir(ta_root_dir + '/local')
    src = ta_root_dir + "/default/props.conf"
    dst = ta_root_dir + "/local/props.conf"
    shutil.copyfile(src,dst)

    f = ta_root_dir + "/local/props.conf"
    if agentmode == "python":
        with open(f,"w") as conf:
            stanza = "[source::.../*.nmon]\n"
            conf.write(stanza)
            if parsingmode in ('json'):
                stanza = "unarchive_cmd = $SPLUNK_HOME/etc/apps/" + ta_root_dir + "/bin/nmon2csv.py --json_output\n"
            else:
                stanza = "unarchive_cmd = $SPLUNK_HOME/etc/apps/" + ta_root_dir + "/bin/nmon2csv.py\n"
            conf.write(stanza)
            conf.write("\n")
            stanza = "[source::.../*.nmon.gz]\n"
            conf.write(stanza)
            stanza = "unarchive_cmd = $SPLUNK_HOME/etc/apps/" + ta_root_dir + "/bin/nmon2csv.py --json_output\n"
            conf.write(stanza)
    elif agentmode == "perl":
        with open(f,"w") as conf:
            stanza = "[source::.../*.nmon]\n"
            conf.write(stanza)
            stanza = "unarchive_cmd = $SPLUNK_HOME/etc/apps/" + ta_root_dir + "/bin/nmon2csv.pl\n"
            conf.write(stanza)
            conf.write("\n")
            stanza = "[source::.../*.nmon.gz]\n"
            conf.write(stanza)
            stanza = "unarchive_cmd = $SPLUNK_HOME/etc/apps/" + ta_root_dir + "/bin/nmon2csv.pl\n"
            conf.write(stanza)

elif agentmode in ('shell') and parsingmode in ('json'):
    os.mkdir(ta_root_dir + '/local')
    src = ta_root_dir + "/default/props.conf"
    dst = ta_root_dir + "/local/props.conf"
    shutil.copyfile(src,dst)

    f = ta_root_dir + "/local/props.conf"

    with open(f, "w") as conf:
        stanza = "[source::.../*.nmon]\n"
        conf.write(stanza)
        stanza = "unarchive_cmd = $SPLUNK_HOME/etc/apps/" + ta_root_dir + "/bin/nmon2csv.sh --json_output\n"
        conf.write(stanza)
        conf.write("\n")
        stanza = "[source::.../*.nmon.gz]\n"
        conf.write(stanza)
        stanza = "unarchive_cmd = $SPLUNK_HOME/etc/apps/" + ta_root_dir + "/bin/nmon2csv.sh --json_output\n"
        conf.write(stanza)

# Manage options for nmon.conf
if parsingmode in ('json'):
    src = ta_root_dir + "/default/nmon.conf"
    dst = ta_root_dir + "/local/nmon.conf"
    shutil.copyfile(src, dst)
    f = ta_root_dir + "/local/nmon.conf"
    with open(f, "w") as conf:
        stanza = "\n# Custom options for the fifo_consumer.sh\n"
        conf.write(stanza)
        stanza = "nmon2csv_options=\"--mode realtime" + " --json_output\"\n"
        conf.write(stanza)
        conf.write("\n")

# Don't use "with" statement in tar creation for Python 2.6 backward compatibility
tar_file = ta_root_dir + '.tgz'
out = tarfile.open(tar_file, mode='w:gz')

try:
    out.add(ta_root_dir)
finally:
    msg = 'INFO: ************* Tar creation done of: ' + tar_file + ' *************'
    print (msg)
    out.close()

# remove Agent directory
if os.path.isdir(ta_root_dir):
        shutil.rmtree(ta_root_dir)

print ('\n*** Agent Creation terminated: To install the agent: ***\n')
print (' - Upload the tgz Archive ' + tar_file + ' to your Splunk deployment server')
print (' - Extract the content of the TA package in $SPLUNK_HOME/etc/deployment-apps/')
print (' - Configure the Application (set splunkd to restart), server class and associated clients to push the new'
       ' package to your clients\n')

# END
print ('Operation terminated.\n')
sys.exit(0)
